### DISCUSSION TITLE
Feature Request: Hellman Attacks

### DISCUSSION THREAD

**Discussion by jwest75674** [madMAx43v3r/chia-plotter#371] [General]
As you know, your chia-plotter does a fantastic job of utilizing system resources more fully, allows plotting to occur at, or near, a machine's capabilities without needing to tweak plot times or delays, while also reducing nvme TBW.

I consider these to be major improvements the make the plotting process easier, less NVME drive-consuming, and (particularly for users with poorly optimized parallelized plotting,) an increased number of plots per day. _i.e. faster_

While I saw a comment you by @madMAx43v3r considering the exploration of openGL to further increase plot speed, (which would be awesome,) I wonder if the [Hellman Attack referenced in chiapos](https://github.com/Chia-Network/chiapos/tree/main/hellman_example) has been considered for this plotter?

I'll be forward with the fact that I do not totally understand the Hellman Attack, [the related PDF](https://github.com/Chia-Network/chiapos/blob/main/hellman_example/Hellman%20attacks.pdf), or [the LASEC paper](https://lasec.epfl.ch/pub/lasec/doc/Oech03.pdf). This is one of two reasons I am writing this instead of working on it directly. _The second being I am not yet proficient with C. :'(_ .

My understanding is that the result of processing would be smaller final plot sizes at the same k32, in exchange for extra computation, computation time, memory use. In other words, more efficient use of storage space, or increased likelyhood to win chia rewards for a given amount of storage space.

Alongside, this would be true to the stated goal of Chia, to be an eco friendly alternative to feeding more energy to GPUs (unlike using openGL?).

While [this functionality doens't appear to be on the roadmap for chiapos currently](https://github.com/Chia-Network/chiapos/pull/192), even if it was, I suspect it would greatly benefit from the threaded implementation used here.

The way I see it, the fantastic work already complete allows the plotting process to be:

- Easier (for many individuals)
- Faster (for poorly optimized plotting setups, or on a per plot basis)
- Less drive-consuming (reducing the number of NVME drives destroyed over time)

Succesfully implementing Hellman Attack processing would mean that maxMAx43v3r/chia-plotter would then also:

- Be more storage efficient (Reduced HDD costs over time / greater likelyhood to win for given storage)
- Allow for (small/med) farmers to continue to remain engaged in the community once their existing drives are full (as many find themselves currently)
- Contribute a whole new functionality/paradime to Chia.
- Potentially more true to the hope of reducing environmental impact (though philisophical and up for debate)

**Comment by TipTaco** [comment]
While the Hellman Attack may produce plot files that are smaller, it comes at the cost of more computation to generate the plot as well as more computation when responding to challenges for that plot. The more space optimised a plot is, the more computation is required to fill in ([by brute force](https://www.chia.net/assets/Chia_Proof_of_Space_Construction_v1.1.pdf#page=25&zoom=100,90,314)) the missing values when responding to a challenge or submitting a proof, and so this would make it less eco friendly to farm.

In addition to more compute power required, it is likely that more disk IO would be needed during farming. Some portion of the Rainbow table (a required part of the Hellman Attack) must be read from disk every time a plot or plots pass the filter. For k=32, this may only be tens of MB, but if checked frequently enough, it could lead to more wear.

I still would be interested if @madMAx43v3r considered a separate post-plotting tool that [performed the Hellman Attack after the fact.](https://github.com/Chia-Network/chiapos/pull/192#issuecomment-846592341) In terms of an individuals profitability, a little more compute and disk IO is probably acceptable for approximately 10% more plots when farming <200TB.

**Comment by cayolblake** [comment]
@jwest75674 where is it mentioned that it reduces TBW?

**Comment by madMAx43v3r** [comment]
hellman attack is not possible, chia has been designed such that it's not profitable.
